src/cargo/core/compiler/mod.rs

```diff
@@ -993,7 +1004,24 @@ fn build_deps_args<'a, 'cfg>(
             v.push(cx.files().out_dir(dep));
             v.push(&path::MAIN_SEPARATOR.to_string());
             v.push(&output.path.file_name().unwrap());
-            cmd.arg("--extern").arg(&v);
+
+            let mut private = false;
```

Comment:
Just from a style standpoint, I would avoid the flag here.  Just have something like (pseudo-code):
```rust
if …feature enabled… && !bcx.is_public_dependency(…) {
    cmd.arg("--extern-private");
    *need_unstable_opts |= true;
} else {
    cmd.arg("--extern");
}
```